### PR TITLE
drivers: gpio: add Ambiq GPIO driver

### PR DESCRIPTION
**Pull request by msierszulski**
This PR adds the Ambiq GPIO driver, as well as instances in the SoC dtsi file, enabling it on the `apollo4p_evb` board.
It also adds support for LEDs and buttons, and adds an overlay for the `gpio_basic_api` test.

With this PR the following samples work on `apollo4p_evb`:
* `samples/basic/blinky`.
* `samples/basic/threads`.
* `samples/basic/button`.

And also the following tests pass:
* `tests/drivers/gpio/gpio_api_1pin`.
* `tests/drivers/gpio/gpio_get_direction`.
* `tests/drivers/gpio/gpio_basic_api`.

### CODE REVIEW

File: dts/arm/ambiq/ambiq_apollo4p.dtsi
Comment (gmarull) [inline dts/arm/ambiq/ambiq_apollo4p.dtsi:96]: this looks weird, why is pinctrl a child? shouldn't it be the top node as in e.g. stm32?

File: drivers/gpio/gpio_ambiq.c
Comment (gmarull) [inline drivers/gpio/gpio_ambiq.c:305]: is this always required?

File: drivers/gpio/gpio_ambiq.c
Comment (gmarull) [inline drivers/gpio/gpio_ambiq.c:233]: spinlocking in a ISR?

File: drivers/gpio/gpio_ambiq.c
Comment (gmarull) [inline drivers/gpio/gpio_ambiq.c:327]:
```suggestion
#define AMBIQ_GPIO_DEFINE(n)                                                                         \
```

File: drivers/gpio/gpio_ambiq.c
Comment (gmarull) [inline drivers/gpio/gpio_ambiq.c:18]: you're likely looking for `gpio-controller.yaml` `ngpios` instead?

Comment (msierszulski) [reply drivers/gpio/gpio_ambiq.c]: Exactly. Thanks!

File: drivers/gpio/gpio_ambiq.c
Comment (msierszulski) [reply drivers/gpio/gpio_ambiq.c]: I changed it.

File: drivers/gpio/gpio_ambiq.c
Comment (msierszulski) [reply drivers/gpio/gpio_ambiq.c]: I removed it.

File: drivers/gpio/gpio_ambiq.c
Comment (msierszulski) [reply drivers/gpio/gpio_ambiq.c]: I moved it to `ambiq_gpio_pin_interrupt_configure`.

File: dts/arm/ambiq/ambiq_apollo4p.dtsi
Comment (msierszulski) [reply dts/arm/ambiq/ambiq_apollo4p.dtsi]: I changed it, so now the top node is `pinctrl`.

File: dts/arm/ambiq/ambiq_apollo4p.dtsi
Comment (gmarull) [inline dts/arm/ambiq/ambiq_apollo4p.dtsi:108]: is pin-offset representing an offset from the pinctrl reg entry? if so that looks wrong @mbolivar-ampere @galak

**Review by mbolivar-ampere — DISMISSED** [review]
@gmarull if you can't answer your question from the description in the binding for this compatible, then that's a problem that needs to be fixed first. The description in YAML should be clear enough to understand without having to ask for clarification in the PR.

Once we know what this is trying to do, we can decide if it's appropriate to put here.

**Comment by msierszulski** [comment]
The answer to why we use this `pin-offset` is that the Ambiq GPIO is actually a single instance with 128 pins. From what I have seen during the implementation of the driver, the Zephyr GPIO API is only able to handle 32 pins per instance, for example `gpio_port_pins_t`, which is actually `uint32_t` with the pin set to index `n` (`1 << n`). We calculate the correct pin number by adding the appropriate pin-offset to the pins passed from the API. So this is a kind of workaround to make it work with this API.
So @gmarull, @mbolivar-ampere, is this acceptable? If yes, I will provide a better explanation of the description in YAML. If not, what can we change to make it better?

**Comment by gmarull** [comment]
> The answer to why we use this `pin-offset` is that the Ambiq GPIO is actually a single instance with 128 pins. From what I have seen during the implementation of the driver, the Zephyr GPIO API is only able to handle 32 pins per instance, for example `gpio_port_pins_t`, which is actually `uint32_t` with the pin set to index `n` (`1 << n`). We calculate the correct pin number by adding the appropriate pin-offset to the pins passed from the API. So this is a kind of workaround to make it work with this API.
> 
> So @gmarull, @mbolivar-ampere, is this acceptable? If yes, I will provide a better explanation of the description in YAML. If not, what can we change to make it better?

Provided this seems like a limitation of the API, difficult to solve (moving to uint64_t would not even be an option...) I guess it is acceptable to create pseudo-nodes... I still think you should just use the reg property to specify the _offset_, ie,

```
#address-cells = <1>;
#size-cells = <0>;

gpioa: gpio-controller@0 {
    reg = <0>;
    ...
```

I will leave the final decision to DT maintainers here, though.

**Comment by henrikbrixandersen** [comment]
> I'm reluctant to allow a devicetree change just to work around a limitation of the GPIO API.
> 
> @henrikbrixandersen any comment on the following?
> 
> > The answer to why we use this pin-offset is that the Ambiq GPIO is actually a single instance with 128 pins. From what I have seen during the implementation of the driver, the Zephyr GPIO API is only able to handle 32 pins per instance, for example gpio_port_pins_t, which is actually uint32_t with the pin set to index n (1 << n). We calculate the correct pin number by adding the appropriate pin-offset to the pins passed from the API. So this is a kind of workaround to make it work with this API.
> 
> Could we handle this in the GPIO API instead?

We can hide the limitation of the 32 bits/port using GPIO maps in devicetree. An example of this can be found here: https://github.com/zephyrproject-rtos/zephyr/blob/697526204764acdaae6ae1e9ab8862c82ff557d5/dts/riscv/neorv32.dtsi#L88

It should be possible to derive the offset from the child nodes.

**Comment by gmarull** [comment]
> > I'm reluctant to allow a devicetree change just to work around a limitation of the GPIO API.
> > @henrikbrixandersen any comment on the following?
> > > The answer to why we use this pin-offset is that the Ambiq GPIO is actually a single instance with 128 pins. From what I have seen during the implementation of the driver, the Zephyr GPIO API is only able to handle 32 pins per instance, for example gpio_port_pins_t, which is actually uint32_t with the pin set to index n (1 << n). We calculate the correct pin number by adding the appropriate pin-offset to the pins passed from the API. So this is a kind of workaround to make it work with this API.
> > 
> > 
> > Could we handle this in the GPIO API instead?
> 
> We can hide the limitation of the 32 bits/port using GPIO maps in devicetree. An example of this can be found here:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/697526204764acdaae6ae1e9ab8862c82ff557d5/dts/riscv/neorv32.dtsi#L88
> 
> It should be possible to derive the offset from the child nodes.

While this could be a solution, this still modifies hardware description to overcome an API (software) limitation.

**Comment by gmarull** [comment]
> I'm reluctant to allow a devicetree change just to work around a limitation of the GPIO API.

+1, note that we have a few though, like creating "virtual" subnodes because we have the 1:1 dt node - device limitation.

**Comment by msierszulski** [comment]
Hi @gmarull @mbolivar-ampere. I have implemented the GPIO pin offset as indicated by @gmarull. Please take a look.

I also rebased the code because there were some conflicts.

File: dts/arm/ambiq/ambiq_apollo4p.dtsi
Comment (gmarull) [inline dts/arm/ambiq/ambiq_apollo4p.dtsi:111]: After looking at the HAL (sigh), and its SVD files, I see that hw engineers had fun by offsetting per-register instead of creating multiple independent ports.

```
    <peripheral>
      <name>FPIO</name>
      <version>1.0</version>
      <description>Fast PIO access</description>
      <!-- <groupName>GROUP_FPIO</groupName> -->
      <baseAddress>0x48001000</baseAddress>
      <size>32</size>
      <access>read-write</access>

      <addressBlock>
        <offset>0</offset>
        <size>0x00000070</size>
        <usage>registers</usage>
      </addressBlock>

      <registers>
        <register>
          <name>RD0</name>
          <description>GPIO Input 0 (31-0)</description>
          <addressOffset>0x00000000</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD0</name>
              <description>GPIO31-0 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD1</name>
          <description>GPIO Input 1 (63-32)</description>
          <addressOffset>0x00000004</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD1</name>
              <description>GPIO63-32 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD2</name>
          <description>GPIO Input 2 (95-64)</description>
          <addressOffset>0x00000008</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD2</name>
              <description>GPIO95-64 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>RD3</name>
          <description>GPIO Input 3 (127-96)</description>
          <addressOffset>0x0000000C</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>RD3</name>
              <description>GPIO127-96 Reads pin state - read only. Returns the pad pin state for pins 0-31 if the PINCFG's input enable (INPEN) is active and RDZERO is inactive.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
        <register>
          <name>WT0</name>
          <description>GPIO Output 0 (31-0)</description>
          <addressOffset>0x00000010</addressOffset>
          <size>32</size>
          <access>read-write</access>
          <resetValue>0x00000000</resetValue>
          <resetMask>0xFFFFFFFF</resetMask>

          <fields>
            <field>
              <name>WT0</name>
              <description>GPIO31-0 Reads or writes pin state. Writes of 1 bits set output pad signal if the GPIO is enabled for output. Reads return status, including sets/clears through the WTS and WTC registers.</description>
              <bitRange>[31:0]</bitRange>
              <access>read-write</access>

            </field>
          </fields>
        </register>
```

and in the HAL you can clearly see it, e.g.

```
#define AM_HAL_GPIO_RDn(pin)    ((volatile uint32_t *)&GPIO->RD0  + (((pin) >> 5) & 0x3))
```

So technically speaking, one has 1 port and 32*4 pins. I wonder then why we have `gpioa`, `gpiob`, etc in the devicetree. I guess Ambiq users will find this weird, I can't see any gpioa/b/etc in the sample code they have.

Comment (msierszulski) [reply dts/arm/ambiq/ambiq_apollo4p.dtsi:111]: I don't really understand what you mean by that. As I said:
>  The answer to why we use this pin offset is that the Ambiq GPIO is actually a single instance with 128 pins.

The whole discussion was about this limitation. Can you please let me know if the current workaround the API limitation is OK with you, and if not - how should I approach it differently?

Comment (RichardSWheatley) [reply dts/arm/ambiq/ambiq_apollo4p.dtsi:200]: @msierszulski Ambiq has done testing with the code from this pull request.
It works with the apollo4p_evb and looks to be inline with the rest of the Zephyr ecosystem from our point of view.
Thanks for the good work!

**Comment by tgorochowik** [comment]
@gmarull could you please take a look if the current implementation (which follows your original suggestion) is acceptable? If not we're happy to rework this a bit further, but it'd be nice to reach an agreement before changing the implementation again.

**Comment by gmarull** [comment]
> @gmarull could you please take a look if the current implementation (which follows your original suggestion) is acceptable? If not we're happy to rework this a bit further, but it'd be nice to reach an agreement before changing the implementation again.

I think introducing artificial names (ie gpioa, b, c...) is not a good idea after all because these don't exist in the reference manual as far as I can see. I don't have a solution to this problem, so I'll leave it to other reviewers to suggest something.

**Comment by mateusz-holenko** [comment]
@gmarull As suggested, we've changed the naming scheme of gpio ports to better reflect the actual hardware. Due to limitations in Zephyr we still had to split the single controller into separate nodes, but now we use actual pin numbers so we hope this is more readable.

File: tests/drivers/gpio/gpio_basic_api/boards/apollo4p_evb.overlay
Comment (RichardSWheatley) [inline tests/drivers/gpio/gpio_basic_api/boards/apollo4p_evb.overlay:8]: Is there a reason the test only covers the first set of GPIO pins and not the whole set of coverage?

Comment (msierszulski) [reply tests/drivers/gpio/gpio_basic_api/boards/apollo4p_evb.overlay:8]: This test is called `GPIO 2-Pin Test` and does not support testing more than 2 pins. 
There is also a need to use the same gpio controller: 
https://github.com/zephyrproject-rtos/zephyr/blob/fb003d1aa80cac2db7ae87c83f862760f5c8e895/tests/drivers/gpio/gpio_basic_api/src/main.c#L23-L33
We can change the `gpio0_31` controller to another, but the results will be exactly the same.

**Comment by msierszulski** [comment]
Hi @gmarull @mbolivar-ampere!
I have rebased the code due to some conflicts. I also added the connector header definition and added its usage in the `gpio_basic_api` test. This connector is needed to properly handle the connection to the shield (https://github.com/zephyrproject-rtos/zephyr/pull/61895).
Can you please take another look at this?

**Comment by gmarull** [comment]
> Hi @gmarull @mbolivar-ampere! I have rebased the code due to some conflicts. I also added the connector header definition and added its usage in the `gpio_basic_api` test. This connector is needed to properly handle the connection to the shield (#61895). Can you please take another look at this?

I'll leave the decision to dt/gpio maintainers

**Comment by MaureenHelm** [comment]
@mbolivar-ampere please revisit

**Comment by mbolivar-ampere** [comment]
> @mbolivar-ampere please revisit

I think we are waiting for @henrikbrixandersen to confirm if his comment in https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1679490658 has been addressed

**Comment by RichardSWheatley** [comment]
> > @mbolivar-ampere please revisit
> 
> I think we are waiting for @henrikbrixandersen to confirm if his comment in [#61157 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1679490658) has been addressed

If you look at dts/arm/ambiq/ambiq_apollo4p.dtsi, it appears that his comment has been addressed. We, Ambiq, agree iwth his implementation of that specifically. Is there any reason to continue to wait?

**Comment by mbolivar-ampere** [comment]
> We, Ambiq, agree with this implementation of that specifically. Is there any reason to continue to wait?

Seeing as there are currently no -1's on the PR, it looks like you just need two +1's including one from @fkokosinski (the assignee)

**Comment by henrikbrixandersen** [comment]
> > We, Ambiq, agree with this implementation of that specifically. Is there any reason to continue to wait?
> 
> 
> 
> Seeing as there are currently no -1's on the PR, it looks like you just need two +1's including one from @fkokosinski (the assignee)

But please be aware that we are currently in feature freeze for v3.5.0.

**Comment by tgorochowik** [comment]
Rebased to resolve conflicts

**Comment by aaronyegx** [comment]
Just a soft reminder, there's conflict again in the branch, and we need to resolve it first. @tgorochowik @msierszulski

**Comment by tgorochowik** [comment]
Rebased again to solve conflicts

**Comment by aaronyegx** [comment]
@tgorochowik , there are conflicts again. We might your help to coordinate to move forward the long-delayed PR.

File: dts/bindings/gpio/ambiq,gpio.yaml
Comment (henrikbrixandersen) [inline dts/bindings/gpio/ambiq,gpio.yaml:13]: This feels like a slight misuse of the `reg` property, which has well-defined meaning.

File: dts/bindings/gpio/ambiq-header.yaml
Comment (henrikbrixandersen) [inline dts/bindings/gpio/ambiq-header.yaml:5]: This seems very specific to that one board. Do you plan to add any shields matching this header, or why is the binding here?

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]

File: dts/bindings/gpio/ambiq,gpio.yaml
Comment (msierszulski) [reply dts/bindings/gpio/ambiq,gpio.yaml:13]: This was originally changed because of https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1677318814 and https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1680156614 Can you take another look at the whole conversation?

File: dts/bindings/gpio/ambiq-header.yaml
Comment (msierszulski) [reply dts/bindings/gpio/ambiq-header.yaml:5]: As mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/61157#issuecomment-1719012866, we plan to add a new shield in https://github.com/zephyrproject-rtos/zephyr/pull/61895. We will reopen this PR after merging this one.

**Comment by aaronyegx** [comment]
Hi @msierszulski, there is some conflict again. By the way, this PR includes the general GPIO driver and the Ambiq header definition which will be needed for shield boards. Not sure if we need to split the drivers to multiple PR to speed up the merging for the changes which are not required to change.

**Comment by RichardSWheatley** [comment]
> Hi @msierszulski, there is some conflict again. By the way, this PR includes the general GPIO driver and the Ambiq header definition which will be needed for shield boards. Not sure if we need to split the drivers to multiple PR to speed up the merging for the changes which are not required to change.

The GPIO PR by itself is over 6 months old.
The Display Shield PR has been waiting on this PR.

Splitting them up did not speed up the process.

@henrikbrixandersen  
What is the issue with this PR moving forward as this is holding up Ambiq from adding to the Zephyr ecosystem.

**Comment by RichardSWheatley** [comment]
> > We created another PR #65900 to only include the general GPIO driver for Ambiq platforms and was proposing another node binding solution as suggested to try to move forward this thread. Hi @henrikbrixandersen , could you help review which solution will be more acceptable? We urgently need to know the direction, thanks for your understanding.
> 
> Let's move forward with #65900 and use separate PRs for connectors etc. I will follow up on the latest changes on that PR today.
> 
> In general, best advise I can give, is to use small, incremental PRs when adding new stuff. These are much easier to review and will generally pass through the submission process faster.

Then let's delete this one to remove confusion.

**Pull request closed without merging**
